Correlation guide first draft (#10296)
DataDog/documentation: content/en/logs/guide/ease-troubleshooting-with-cross-product-correlation.md

   * [Correlate application logs](#correlate-application-logs)
   * [Correlate proxy logs](#correlate-proxy-logs)
   * [Correlate database logs](#correlate-database-logs)
2. [Correlate frontend products](#correlate-frontend-products)
   * [Correlate browser logs with RUM](#correlate-browser-logs-with-rum)
3. [Correlate user experience with server behavior](#correlate-user-experience-with-server-behavior)
   * [Correlate RUM views with traces](#correlate-rum-views-with-traces)
   * [Leverage trace correlation to troubleshoot Synthetic tests](#leverage-trace-correlation-to-troubleshoot-synthetic-tests)

**Note**: Depending on your use case, you may skip certain steps below. Steps that are dependent on others are explicitly stated.

## Correlate server-side logs with traces

When your users are encountering errors or high latency in your application, drilling down to view the logs from a problematic request can reveal exactly what went wrong. By pulling together all the logs pertaining to a given request, you can see in rich detail how it was handled from beginning to end so you can quickly diagnose the issue.

> Correlating your logs also eases [aggressive sampling strategy without losing entity-level consistency][2] with the use of attributes like `trace_id` and `session_id`.

Reviewer: ```suggestion
Correlating your logs with traces also eases [aggressive sampling strategy without losing entity-level consistency][2] with the use of attributes like `trace_id` and `session_id`.
```

Author: `Aggressive sampling strategy without losing entity-level consistency` is a benefit of log correlation in general (as a matter of fact, the `session_id` example will come from RUM correlation). I wrote it here because it is the main log correlation, but it is not only with traces

Author: What do you think of duplicating this sentence? Would it be better

In https://docs-staging.datadoghq.com/gturbat/new-guide-log-correlation/logs/guide/ease-troubleshooting-with-cross-product-correlation/?tab=nginx#correlate-server-side-logs-with-traces:

> Correlating your logs with traces also eases [aggressive sampling strategy without losing entity-level consistency][2] with the use `trace_id`.

In https://docs-staging.datadoghq.com/gturbat/new-guide-log-correlation/logs/guide/ease-troubleshooting-with-cross-product-correlation#why-3:

> Correlating your browser logs with RUM also eases [aggressive sampling strategy without losing entity-level consistency][2] with the use `session_id`.

Reviewer: Ah, I see. I misunderstood the relevancy to the overall subject of correlation versus to the header since it's under `Correlate server-side logs with traces`. We can duplicate in both areas I think would be best, just to avoid confusion.

Author: Done in https://github.com/DataDog/documentation/pull/10296/commits/8eecd29c793714c43f68c7f42e635d26e0c4c916

```conf
http {
  log_format main '$remote_addr - $opentracing_context_x_datadog_trace_id $http_x_forwarded_user [$time_local] "$request" '
          '$status $body_bytes_sent "$http_referer" '
          '"$http_user_agent" "$http_x_forwarded_for" ';

  access_log /var/log/nginx/access.log;
}
```

##### Parse trace ID in pipelines

1. Clone the NGINX pipeline.

2. Customize the first [grok parser][2]:
   - In **Parsing rules**, replace the first parsing rule with:
   ```text
   access.common %{_client_ip} %{_ident} %{_trace_id} %{_auth} \[%{_date_access}\] "(?>%{_method} |)%{_url}(?> %{_version}|)" %{_status_code} (?>%{_bytes_written}|-)
   ```
   - In **Advanced settings** under **Helper Rules**, add the line:
   ```text
   _trace_id %{notSpace:dd.trace_id:nullIf("-")}
   ```

3. Add a [trace ID remapper][3] on `dd.trace_id` attribute.

[1]:/tracing/setup_overview/proxy_setup/?tab=nginx
[2]:/logs/processing/processors/?tab=ui#grok-parser
[3]:/logs/processing/processors/?tab=ui#trace-remapper
{{% /tab %}}
{{< /tabs >}}

### Correlate database logs

#### Why?

Database logs are often hard to contextualize due to query similarities, variable anonymization, and high usage.

> For example, production slow queries are hard to reproduce and analyze without investing a lot of time and resources. Let's correlate slow query analysis with traces.

Reviewer: ```suggestion
For example, production slow queries are hard to reproduce and analyze without investing a lot of time and resources. Below is an example how to correlate slow query analysis with traces.
```

        return statement, parameters
```

**Note**: This only correlates logs that include a query statement. Error logs like `ERROR:  duplicate key value violates unique constraint "<TABLE_KEY>"` stay out of context. Most of the time you can still get error information through your application logs.

Clone and customize the PostgreSQL pipeline:

1. Add a new [grok parser][4]:
   ```text
   extract_trace %{data}\s+--\s+dd.trace_id=<%{notSpace:dd.trace_id}>\s+%{data}
   ```

2. Add a [trace ID remapper][5] on `dd.trace_id` attribute.

Here is an example of a slow query execution plan from a slow trace:

{{< img src="logs/guide/ease-troubleshooting-with-cross-product-correlation/slow-query-root-cause.png" alt="Slow query logs correlation" style="width:80%;" >}}

[1]: /integrations/postgres/?tab=host#log-collection
[2]: https://www.postgresql.org/docs/13/auto-explain.html
[3]: https://www.postgresql.org/docs/13/sql-syntax-lexical.html#SQL-SYNTAX-COMMENTS
[4]: /logs/processing/processors/?tab=ui#grok-parser
[5]: /logs/processing/processors/?tab=ui#trace-remapper
{{% /tab %}}
{{< /tabs >}}

## Correlate frontend products

### Correlate browser logs with RUM

#### Why?

> [Browser logs][9] inside a RUM event gives context and insight into causes and consequences. As in the example below, browser logs indicate that the bad query root cause is an invalid user ID.

Reviewer: ```suggestion
[Browser logs][9] inside a RUM event give context and insight into an issue. As in the example below, browser logs indicate that the bad query root cause is an invalid user ID.
```

{{< img src="logs/guide/ease-troubleshooting-with-cross-product-correlation/browser-logs-in-rum.png" alt="Browser logs in a RUM action" style="width:80%;" >}}

#### How?

Browser logs and RUM events are automatically correlated as explained in the [RUM billing FAQ][10]. [Matching configuration between RUM and logs SDK][11] is required.

## Correlate user experience with server behavior

> Traditional backend and frontend monitoring are siloed and require separate workflows to troubleshoot across the stack. Datadog full-stack correlations let you identify root cause (whether it comes from a browser issue or a database downtime) and estimate user impact in a quick and intuitive manner.

Reviewer: ```suggestion
Traditional backend and frontend monitoring are siloed and require separate workflows to troubleshoot across your stack. Datadog full stack correlations lets you identify a root cause (whether it comes from a browser issue or a database downtime) and estimate user impact.
```

> This section walks you through the steps to enable such correlations:

Reviewer: ```suggestion
This section walks you through the steps to enable these types of correlations:
```
